### ISSUE TITLE
Code Style Consistency

### ISSUE THREAD

**Issue by TkTech** [lemire/simdjson#104]
The code currently uses an odd mix of indentation patterns, function naming (isValid, move_forward), comments, etc...

Would there be any objection to normalizing the style? As you get more external contributors this will become more important to help keep the code maintainable. A new test would double check the style before it passes.

https://clang.llvm.org/docs/ClangFormat.html for example is a very useful tool for doing so automatically, ex: `clang-format -style=google` which would apply [Google's style guide](https://google.github.io/styleguide/cppguide.html).

**Comment by geofflangdale** [comment]
This is a good idea. I wouldn't necessarily want to immediately take on every element of the Goog's style guide blindly - I can see a bunch of churn for no good reason. But we haven't really figured out a style guide for stuff like indenting and identifiers and it makes sense to bring in something else rather than try to dream up our own style guide (there's 3 months we won't get back!).

I like Google's recommendations for most things. I'm having trouble getting used to 2 spaces rather than 4, but Daniel's style is already 2, and I don't have strong feelings.

How aggressive will ClangFormat be, exactly? Does it fix dodgy identifiers or just flag them?

**Comment by lemire** [comment]
I am super annoyed at that myself... so yeah, I am supporter of something nicer.

I have resisted applying clang-format to the project but I think we would welcome both a reformatting *and* maybe some tooling to check formatting (e.g., during unit tests). I am against adding clang-format as a dependency for developers, because dependencies are evil, but it would be nice to know when a PR breaks the format.

This being said, one has to be sure, when doing a reformat PR, that one does not cross other ongoing PRs because it creates conflicts from hell. Care is needed.

**Comment by geofflangdale** [comment]
I've got some big 'smashing' reworks coming which would be a good time to apply clang-format to some pretty large chunks of code. I need to refactor Stage 1 to make it less verbose and somewhat more modular (without breaking performance, which is always delicate). This will change a lot of LOC at once _anyway_. AFAICT there aren't that many people working in that area.

Are we agreed that `clang-format -style=google` is canonical? It sounds OK to me but maybe I should play around with it.

**Comment by lemire** [comment]
@geofflangdale 

I personally don't have any strong feelings regarding formatting. What I think we all agree with is *consistency*.

**Comment by TkTech** [comment]
Sorry if it seemed like I was pushing the Google style guide, I just meant for it to be an example!

As an open-source contributor you get used to working on projects with a wide range of styles. You just adapt to whatever the adjacent code looks like, it's someone else's project after all. But of course that only works if the styling is consistent.

**Comment by geofflangdale** [comment]
I used the google style guide on my refactoring of stage1_find_marks.cpp and liked it.

**Comment by lemire** [comment]
>  it's someone else's project after all

I am pretty sure that I speak for both myself and @geofflangdale when I state that "owning" this project is not a goal.

**Comment by geofflangdale** [comment]
If anyone wants a look at what this looks like, see my recent check-in of stage1_find_marks.cpp - this went through clang-format's idea of the Google style guide, at least.

**Comment by lemire** [comment]
We should fix this soon, hopefully for release 0.2.

cc @ioioioio

**Comment by ioioioio** [comment]
I pushed a branch ([style_uniformization](https://github.com/lemire/simdjson/tree/style_uniformization)) on which I applied a clang-format with the LLVM style. I chose that one because it is the one that gave the result that was the closest to what we already have.

However, I don't think it is the good option to apply LLVM style guide's naming convention (https://llvm.org/docs/CodingStandards.html), since there would have too much things to change. I suggest to stick with the snake_case for functions, variables, structs and enums, sinces it seems to be what have been the most used.

I'll try to do it quick tomorrow.
